File: 0543-diameter-of-binary-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
    if root == nil {
        return 0
    }

    max := 0
    var dfs func(root *TreeNode) int
    dfs = func (root *TreeNode) int {
        if root == nil {
            return 0
        }

        left := dfs(root.Left)
        right := dfs(root.Right)

        if left + right > max {
            max = left + right
        }

        return 1 + maxFunc(left, right)
    }

    dfs(root)
    return max
}

func maxFunc(a,b int) int {
    if a > b {
        return a
    }
    
    return b
}

File: 0543-diameter-of-binary-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_value = -1

        def dfs(root):
            nonlocal max_value
            if root is None:
                return 0

            left = dfs(root.left)
            right = dfs(root.right)
        
            if left + right > max_value:
                max_value = left + right

            return 1 + max(left, right)

        dfs(root)
        return max_value
